### PR TITLE
fix(analytics): logEvent params can be objects, further fix for…

### LINKED ISSUE #2876: 🔥 Cannot send array/object in logEvent in v6
## Issue

In v5 you [could provide arrays / objects](https://rnfirebase.io/docs/v5.x.x/analytics/reference/analytics#logEvent) within the `logEvent` `params` object, and such usage is even provided in [Firebase's own examples](https://firebase.google.com/docs/analytics/events?platform=web#log_events_3).

In v6 `logEvent` is [restricted to having strings, numbers and booleans](https://invertase.io/oss/react-native-firebase/v6/analytics/reference/module#logEvent) as values and it throws an exception if any other type is provided:

Screenshot of error

It seems this is a major limitation in v6 compared to what Firebase analytics offers.

_(Edit: initially wrote getEvent instead of logEvent, still wrong in the title)_

---

### iOS

<details><summary>Click To Expand</summary>
<p>

#### `ios/Podfile`:

- [ ] I'm not using Pods
- [x] I'm using Pods and my Podfile looks like:

```ruby
# N/A
```

#### `AppDelegate.m`:

```objc
// N/A
```

</p>
</details>

---

### Android

<details><summary>Click To Expand</summary>
<p>

#### Have you converted to AndroidX?

- [ ] my application is an AndroidX application?
- [ ] I am using `android/gradle.settings` `jetifier=true` for Android compatibility?
- [ ] I am using the NPM package `jetifier` for react-native compatibility?

#### `android/build.gradle`:

```groovy
// N/A
```

#### `android/app/build.gradle`:

```groovy
// N/A
```

#### `android/settings.gradle`:

```groovy
// N/A
```

#### `MainApplication.java`:

```java
// N/A
```

#### `AndroidManifest.xml`:

```xml
<!-- N/A -->
```

</p>
</details>

---

## Environment

<details><summary>Click To Expand</summary>
<p>

**`react-native info` output:**

```
info
  React Native Environment Info:
    System:
      OS: macOS 10.14.6
      CPU: (12) x64 Intel(R) Core(TM) i7-8850H CPU @ 2.60GHz
      Memory: 149.97 MB / 32.00 GB
      Shell: 5.0.9 - /usr/local/bin/bash
    Binaries:
      Node: 12.8.1 - /usr/local/bin/node
      Yarn: 1.17.3 - /usr/local/bin/yarn
      npm: 6.11.3 - /usr/local/bin/npm
      Watchman: 4.9.0 - /usr/local/bin/watchman
    SDKs:
      iOS SDK:
        Platforms: iOS 13.1, DriverKit 19.0, macOS 10.15, tvOS 13.0, watchOS 6.0
    IDEs:
      Android Studio: 3.5 AI-191.8026.42.35.5791312
      Xcode: 11.1/11A1027 - /usr/bin/xcodebuild
    npmPackages:
      react: 16.9.0 => 16.8.3
      react-native: 0.61.4 => 0.59.9
```

- **Platform that you're experiencing the issue on**:
  - [ ] iOS
  - [ ] Android
  - [ ] **iOS** but have not tested behavior on Android
  - [ ] **Android** but have not tested behavior on iOS
  - [x] Both
- **`react-native-firebase` version you're using that has this issue:**
  - `6.0.3`
- **`Firebase` module(s) you're using that has the issue:**
  - `Analytics`
- **Are you using `TypeScript`?**
  - `Y` & `3.7.2`
  
</p>
</details>

---

Think `react-native-firebase` is great? Please consider supporting all of the project maintainers and contributors by donating via our [Open Collective](https://opencollective.com/react-native-firebase/donate) where all contributors can submit expenses. [[Learn More]](https://invertase.io/oss/react-native-firebase/contributing/donations-expenses)

- 👉 Check out [`React Native Firebase`](https://twitter.com/rnfirebase) and [`Invertase`](https://twitter.com/invertaseio) on Twitter for updates on the library.

### PR DESCRIPTION
**Pull request by plaa**
### Summary

Actually fixes #2876 which was supposedly fixed by #3219.

In #3219 code was included that mangled the `params` object before passing it to the native Firebase libraries.  This code was completely broken, and resulted in all arrays and objects being converted into the fixed string `"[object Object]"`.

Simply removing the mangling code fixes the issue.  This allows passing arrays and objects to the native library (which on Android is recursively converted by `Arguments.toBundle` and on iOS doesn't need any conversion).  All existing code continues to work as-is as they don't include arrays / objects.

In #3219 types were forgotten to be updated.  65ec7eb updated the TypeScript types.  This PR also updates the Flow types.

### Checklist

- [x] Supports `Android`
- [x] Supports `iOS`
- [ ] `e2e` tests added or updated in packages/**/e2e
- [x] Flow types updated
- [ ] Typescript types updated

### Test Plan

This change cannot break anything, as it simply removes mangling of objects / arrays which weren't allowed in previous versions anyway.

In our setup using GTM + Ecommerce we were unable to get `view_item` / `add_to_cart` events to include the item details when provided as top-level properties.  Using the patched library we were able to provide them inside the `items` array, which worked properly.  (Google's own documentation on this is inconsistent and in places self-contradictory.)

Before (doesn't work):
```
logEvent('view_item', {
  item_id: '1234',
  item_name: 'Acme Sledgehammer'
})
```

After (works):
```
logEvent('view_item', {
  items: [
    {
      item_id: '1234',
      item_name: 'Acme Sledgehammer'
    }
  ]
})
```

### Release Plan

Same message as designed for #2876 / #3219.

I hope this will be included in the 6.4.0 release, as the analytics is useless for us without it.

### CODE REVIEW

File: packages/analytics/lib/index.js.flow [hunk 1]
```diff
@@ -623,7 +623,7 @@ export interface Module extends ReactNativeFirebaseModule {
    * @param name Event name must not conflict with any Reserved Events. 100 characters is the maximum length for param key names.
    * @param params Parameters to be sent and displayed with the event.
    */
-  logEvent(name: string, params?: { [key: string]: string  | number | boolean } | {}): Promise<void>;
+  logEvent(name: string, params?: { [key: string]: any } | {}): Promise<void>;
```
Comment (Ehesp) [inline packages/analytics/lib/index.js.flow]: The `| {}` part can be removed here I think?

Comment (plaa) [reply packages/analytics/lib/index.js.flow]: Not sure.  It was included also in TypeScript definitions and I don't know Flow that much.  Just noticed that the update was forgotten here as well.

Comment (plaa) [reply packages/analytics/lib/index.js.flow]: My bad, wasn't included in TypeScript.  I removed it.  In any case, the beef of the PR is the code change.

**Comment by Salakar** [comment]
@russellwheatley I swear we updated this to be identical to the Web SDK, thought we saw they were doing `.toString` also? Can't find it in the source though 😅

**Comment by russellwheatley** [comment]
Yeah, I thought we did get it to match the web version as it was showing up as  [Object object] in the console. Not sure if we located it in the web sdk source however. It might be occurring within the console itself I guess. Would need to look into it again as many moons have passed since then. I'll look into it 👍

**Comment by plaa** [comment]
If anyone else is having issues with this, I created a [temporary fork of @react-native-firebase/analytics](https://github.com/plaa/react-native-firebase-analytics-tmp).

Usage in package.json:

```
"@react-native-firebase/analytics": "plaa/react-native-firebase-analytics-tmp#master",
"@react-native-firebase/app": "^6.4.0-rc2",
```

**Comment by russellwheatley** [comment]
hi @plaa, 
I've tried running this code:
```
logEvent('view_item', {
  items: [
    {
      item_id: '1234',
      item_name: 'Acme Sledgehammer'
    }
  ]
})
```
along with your PR changes and the `items` array did not appear in the firebase console or Analytics console. You mention that you were running it as `ecommerce`. I'm unable to enable `ecommerce` when my Google Analytics property is an "App" or "Web and App". As far as I can tell, it is available for  "Web" properties.

Could you please instruct me on how you were able to receive this content within the firebase console so I can confirm, please?

Thank you.

**Comment by plaa** [comment]
Hi,

We have separate professionals who have set up the analytics system so unfortunately I don't know or understand the exact details. We use Firebase analytics + GTM libraries in the client and also use Ecommerce in analysis.

Google's own documentation on how item details should be passed in view_item / add_to_cart events is varying and occasionally self-contradicting.

Android docs now instruct placing the item details in the `items` array as type `Parcelable`, which is implemented by `Bundle` used by react-native-firebase.  Still last week the Android link below instructed placing the details on the top level.  iOS also instructs placing them in the `items` array.
https://firebase.google.com/docs/reference/android/com/google/firebase/analytics/FirebaseAnalytics.Event#VIEW_ITEM
https://firebase.google.com/docs/reference/ios/firebaseanalytics/api/reference/Constants#/c:FIREventNames.h@kFIREventViewItem

We tried putting the item details on the top level (as previously instructed for Android) and in this case the fields were displayed in the Firebase DebugView console, but they did not appear in the further reports.  When we placed them inside the `items` array with the patched library, they do not show up in DebugView but are present in the further reports.  (I have sent different item details for both cases and those sent using the `items` array came through and those without the array didn't.)

Both Android and iOS documentation for `logEvent` states that it only supports string and numeric types, which is clearly wrong in the case of view_item / add_to_cart.  I believe this is the source for the validation that has been added to the react-native-firebase/analytics library in v6.
https://firebase.google.com/docs/reference/android/com/google/firebase/analytics/FirebaseAnalytics#logEvent(java.lang.String,%20android.os.Bundle)
https://firebase.google.com/docs/reference/ios/firebaseanalytics/api/reference/Classes/FIRAnalytics#/c:objc(cs)FIRAnalytics(cm)logEventWithName:parameters:

The react-native-firebase/analytics library is a thin wrapper around the Firebase native libraries.  As of v6, there has been validation that prevents sending parameter types that are not string/numeric.  This PR simply removes that validation, allowing to send any types that you would be able to send the library in a native app.  The JS lib doesn't _need_ to perform validation, and if there are use cases (such as mine) that do not work with the validation, why not remove it?

Thanks.

**Comment by plaa** [comment]
Hi,

Maybe noteworthy that we originally have a web page with analytics, and we're now developing the mobile app and trying the get the same analytics out of it. This might be the source for ecommerce web/web+app/app disparity. But as I said, I don't understand the exact details.

What I know is that with the patched library the analytics work, and the patch doesn't break anything.

**Comment by Salakar** [comment]
This is up on the 6.4.0 release

**Comment by plaa** [comment]
For anyone finding this thread trying to get Firebase + GA/Ecommerce working:

We finally discovered that for Firebase analytics to work, the product details have to be sent in the top-level object, while GA/Ecommerce requires them to be in the `items`-array.  Thus the same data needs to be sent twice in the analytics event for both to work. 🤦‍♀️

So we're now doing:
```
logEvent('view_item', {
  item_id: '1234',
  item_name: 'Acme Sledgehammer',
  items: [
    {
      item_id: '1234',
      item_name: 'Acme Sledgehammer'
    }
  ]
})
```

**Pull request merged**